### PR TITLE
Add User-friendly Attribute Names and Values to Order Line Items Metadata

### PR DESCRIPTION
**Pull request by vedanshujain**
From https://github.com/woocommerce/woocommerce-rest-api/pull/229

> 
> Fixes #9.
> 
> Ref: p91TBi-2Lt-p2
> 
> This will also allow us to fix these issues for the WooCommerce iOS and Android apps:
> 
>     woocommerce/woocommerce-android#1286
>     woocommerce/woocommerce-ios#1127
>     woocommerce/woocommerce-ios#2280
> 
> Background
> 
> For both the iOS and Android apps, we regularly receive feedback from users that they could not view the details of the variable products in the orders. Example: C013AAPA4G0-p1596070344020400-slack. To find a solution, @AmandaRiu investigated what the iOS and Android apps can do using the REST API and posted her findings at p91TBi-2Lt-p2.
> 
> The apps use the line_items name in the GET /v3/orders response. However, the name does not include all the selected variations. Consider an order for a variable product with color "Blue" and size "Medium". The API endpoint would return something like this (simplified result):
> 
>     "line_items": [
>         {
>             "id": 5,
>             "name": "Variable Hoodie - Blue",
> 
>             "meta_data": [
>                 {
>                     "id": 51,
>                     "key": "pa_store-level-size",
>                     "value": "medium",
>                 },
>                 {
>                     "id": 52,
>                     "key": "product-level-color",
>                     "value": "Blue",
>                 },
>             ],
>     
>         }
> 
> As shown above, the size was not included in the name. This is how it would look like in the iOS app:
> 
> WCAdmin shows the attribute that was omitted:
> 
> Why Not Use Metadata?
> 
> We could use the meta_data array to find the missing attributes. However, this may be unreliable because:
> 
>     We have to analyze name to figure out which attribute is already filled in. This can include finding the dash separator, which I think can be customized too.
>     The value in meta_data is the sanitized value. In the example above, "medium" should be "Medium" (capital M).
> 
> Solution
> 
> The name not showing all the attributes is probably attributed to the generate_product_title function in WooCommerce. It's probably possible to solve the issue by making sure that all attributes are included. However, this won't solve one of our other design requirements (woocommerce/woocommerce-ios#2280) which is to present the attributes separately:
> 
> We would still have to do a string operation to split out the attribute values.
> 
> In this PR, I'm proposing adding the displayable attribute name and values in the meta_data instead:
> 
>     "line_items": [
>         {
>             "id": 5,
>             "name": "Variable Hoodie - Blue",
> 
>             "meta_data": [
>                 {
>                     "id": 51,
>                     "key": "pa_store-level-size",
>                     "value": "medium",
>                     "display_key": "Store Level Size",
>                     "display_value": "Medium"
>                 },
>                 {
>                     "id": 52,
>                     "key": "product-level-color",
>                     "value": "Blue",
>                     "display_key": "Product Level Color",
>                     "display_value": "Blue"
>                 },
>             ],
>     
>         }
> 
> With these available properties, the API clients (e.g. iOS and Android apps) would be able to easily decide how to present the attributes.
> 
> There is still the problem with the name possibly showing one of the attributes. I'm planning to submit another PR to fix that after this one.
> Schema
> 
> I added the new properties above in the schema.
> 
>                             "meta_data": {
>                                 "description": "Meta data.",
>                                 "type": "array",
>                                 "context": [
>                                     "view",
>                                     "edit"
>                                 ],
>                                 "items": {
>                                     "type": "object",
>                                     "properties": {
>                                         "id": {
>                                             "description": "Meta ID.",
>                                             "type": "integer",
>                                             "context": [
>                                                 "view",
>                                                 "edit"
>                                             ],
>                                             "readonly": true
>                                         },
>                                         "key": {
>                                             "description": "Meta key.",
>                                             "type": "string",
>                                             "context": [
>                                                 "view",
>                                                 "edit"
>                                             ]
>                                         },
>                                         "value": {
>                                             "description": "Meta value.",
>                                             "type": "mixed",
>                                             "context": [
>                                                 "view",
>                                                 "edit"
>                                             ]
>                                         },
>                                         "display_key": {
>                                             "description": "Meta key for UI display.",
>                                             "type": "string",
>                                             "context": [
>                                                 "view",
>                                                 "edit"
>                                             ]
>                                         },
>                                         "display_value": {
>                                             "description": "Meta value for UI display.",
>                                             "type": "string",
>                                             "context": [
>                                                 "view",
>                                                 "edit"
>                                             ]
>                                         }
>                                     }
>                                 }
>                             },
> 
> Changes
> 
> These new properties are only applied to Orders API V3 since we are hoping to get fixed in the mobile apps soon. I did this by overriding the get_order_item_data in Controllers\Version3\ WC_REST_Orders_Controller.
> 
> Please let me know if I should include this in V4 as well.
> Testing
> 
> This is just a suggestion but please feel free to test other scenarios. 🙂
> 
>     In wp-admin → Products → Attributes, add a new attribute and add terms to that attribute.
>     Add a new Variable product in wp-admin → Products → Add New.
>     Set this product as Variable.
>     In the product's Attributes section, add the attribute created in Step 1. Set this as "Used for variations". Add values.
>     Add a Custom product attribute. Set this as "Used for variations". Add values.
>     Save the attributes.
>     In the product's Variations section, create variations for all the options. Or create a few variations only.
>     Set prices for all the variations.
>     Publish the product.
>     As a customer, place an order of the created Variable product. Make note of the Order ID.
>     In the API, run GET /wp-json/wc/v3/orders/{OrderID}.
>     Confirm that the line_items meta_data has the display_key and display_value properties and they are correct.
> 
> Unit Test
> 
> I have also added a unit test that I think covers the manual testing scenario above.
>

### CODE REVIEW

**Review by vedanshujain — COMMENTED** [review]
Note that I have already approved the commits in original PR once, so only one review is needed.

**Comment by vedanshujain** [comment]
cc @shiki

**Comment by shiki** [comment]
@vedanshujain I have submitted a PR to fix the failing unit test. → https://github.com/woocommerce/woocommerce/pull/27559

**Comment by astralbodies** [comment]
@vedanshujain Any chance this might be reviewed soonish and merged? Trying to establish a timeline when we can tell users to expect a resolution to the fulfillment issues in the mobile app. Thanks! 😃

**Comment by shiki** [comment]
Hi @claudiosanches! Let me know if there's anything I can help to get this merged. 🙂 I'm not quite sure what the Travis failures are. Are they coding style issues?

**Comment by vedanshujain** [comment]
Looks like the test is failing because the formatting function is excluding non-scalar entities (an array in our case). This a bug that needs to be fixed before we merge, I am working on the fix, should have it ready by tomorrow.

**Comment by shiki** [comment]
@vedanshujain Ah, I see. I believe that's the one I fixed in this PR → https://github.com/woocommerce/woocommerce/pull/27559.

**Comment by vedanshujain** [comment]
LGTM now asked for another review. Marking as a high impact change since it changes the returned data format of a nested property.

File: tests/legacy/unit-tests/rest-api/Tests/Version3/orders.php
Comment (peterfabian) [inline tests/legacy/unit-tests/rest-api/Tests/Version3/orders.php:197]: Would it make sense to use variables to check for `$color_meta_data['display_key']` and `$color_meta_data['display_value']`?

**Review by peterfabian — DISMISSED** [review]
Works fine, tests are passing now, thanks folks, just a small nitpick, but not a blocker.

I tested v2 and v3 using Postman with some funky attributes and it seems to work fine: 
Screenshot 2020-10-14 at 13 18 30


I assume adding new fields should not be a big problem for bw compatibility, so should be good to go I think.

File: tests/legacy/unit-tests/rest-api/Tests/Version3/orders.php
Comment (vedanshujain) [reply tests/legacy/unit-tests/rest-api/Tests/Version3/orders.php]: Added in https://github.com/woocommerce/woocommerce/pull/27492/commits/a332b399d3501ad70f1c761f11a7a2b28310ee44. I will merge when tests pass.

**Comment by vedanshujain** [comment]
Merging as tests passed, the code standard test is failing on all PRs for unrelated reason looks like.

**Pull request merged**
